crates/utf8-converter/src/bitrank.rs

```diff
@@ -0,0 +1,504 @@
+//! A bit-vector data structure, optimized for
+//! [rank](http://bitmagic.io/rank-select.html) operations.
+//!
+//! There is also an opportunistic `select` operation, but the general case has not been
+//! implemented.
+
+type Chunk = u128;
+
+// Static sizing of the various components of the data structure.
+const BITS_PER_BLOCK: usize = 16384;
+const BITS_PER_SUB_BLOCK: usize = 128;
+const SUB_BLOCKS_PER_BLOCK: usize = BITS_PER_BLOCK / BITS_PER_SUB_BLOCK;
+const BITS_PER_CHUNK: usize = 128;
```

Comment:
note: I figured out a potentially faster solution here with a simple trick.
https://github.com/github/rust-gems/pull/37/files#diff-62430fbe79087bea140708a5d90481a016def9872a1180d9217781caa4c83499R113-R133
Ignore all the mapping stuff (it didn't actually reduce the number of page reads).
But you can store the sum up to the middle of a chunk and then either subtract the bits BEFORE the middle or add the bits AFTER the middle. That way, you can use u64 instead of u128 which in my benchmarks was faster.

Maybe something for a separate PR...

Reply:
Yes, this is wild. Do you have those benchmarks lying around anywhere? I can throw them in while I'm here...

Reply:
Sorry, didn't write any for the underlying bitrank stuff...